### PR TITLE
refactor(incrementDay): Use functional `setState`

### PR DESCRIPTION
**Pull request by jeremyckahn**
### What this PR does

This PR refactors `incrementDay` to hopefully avoid race conditions that could conceivably result from stale React state.

### How this change can be validated

ℹ️ This is pretty extensive testing plan and I'm not expecting anyone to actually run through all of the scenarios. A spot check of behavior related to the "End Day" button should be sufficient and appreciated, as what's here is mostly covered by the automated tests.

<details>

<summary>
Scenarios to test
</summary>

* **Starting a New Day (First Day):**
    * [ ] Observe the day counter in the UI change to "Day 1".
    * [ ] See initial game notifications, such as information about the starting loan.
    * [ ] Confirm that if the game is closed and reopened, progress from Day 1 is loaded.
* **Advancing Subsequent Days:**
    * [ ] Observe the day counter in the UI increment by one (e.g., "Day 1" becomes "Day 2").
    * [ ] Confirm that if the game is closed and reopened, progress from the new day is loaded.
* **Online Mode - Market Impact:**
    * [ ] When playing online and ending the day after buying/selling/harvesting, observe notifications about how your actions affected the shared market prices.
    * [ ] Notice that item prices in the shop or when selling may have changed based on these actions.
* **Online Mode - Server Connection Issues:**
    * [ ] If the game fails to connect to the server when ending the day online, see a "Server Error" notification.
    * [ ] Observe that the game switches to offline mode (e.g., online features become unavailable).
* **Offline Mode - Market Changes:**
    * [ ] When playing offline, observe that item prices in the shop or when selling still fluctuate day-to-day (based on local calculations).
* **Automatic Harvesting (Combine Enabled):**
    * [ ] If the combine harvester is enabled, observe mature crops disappear from the field and their harvested items appear in your inventory (or a notification if inventory is full) when the day ends.
* **No Automatic Harvesting (Combine Disabled):**
    * [ ] If the combine harvester is disabled, observe mature crops remain in the field when the day ends.
* **End-of-Day Notifications:**
    * [ ] See relevant notifications for events that occurred "overnight" (e.g., rain, crop growth, financial summaries) appear on screen.
    * [ ] If the game is reloaded, these notifications should reappear if they were not dismissed.

**Field, Crops & Environment:**

* **Crop Growth Progress:**
    * [ ] Observe visual changes in planted crops (e.g., seedlings grow taller, eventually showing mature produce) if they were watered.
    * [ ] If crops were not watered, they should not advance in their growth stage.

**Economy, Events & Winery:**

* **Market Events:**
    * [ ] See a notification that prices for a specific crop have crashed or surged.
* **Market Event: Duration:**
    * [ ] Observe that price crash/surge events last for a set number of days.
    * [ ] See prices return to normal after the event duration expires.
* **Daily Financial Summary:**
    * [ ] Review financial records to see the previous day's revenue and losses correctly logged.
    * [ ] Confirm today's revenue and loss trackers reset for the new day.
* **Market Price Fluctuations (Offline):**
    * [ ] When playing offline, notice that item prices in the shop and sell values still change daily (unless a price event is active).
* **Market Price Fluctuations (Online):**
    * [ ] When playing online, notice that item prices in the shop and sell values change daily, reflecting the shared online market (unless a price event is active).
* **Notification Log Update:**
    * [ ] Check the Farmer's log to see that the previous day's significant events and notifications are recorded.

</details>

### Questions or concerns about this change

I'm not sure of a way to verify that this fix prevents the #546, or is stale state was even the cause of the issue. But this seems like the most likely cause and the best solution available.

### Additional information

- https://discord.com/channels/714539345050075176/822068921653788683/1368728669123776575

### CODE REVIEW

File: src/components/Farmhand/Farmhand.js [hunk 1]
```diff
@@ -1045,87 +1045,126 @@ export default class Farmhand extends FarmhandReducers {
     )
   }
 
-  /*!
-   * @param {boolean} [isFirstDay=false]
+  /**
+   * @param {boolean} isFirstDay
+   * @returns {Promise<string | null>}
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/components/Farmhand/Farmhand.js]: Added in https://github.com/jeremyckahn/farmhand/commit/69bb4d18be7964a18da2e16718a2d8026f782b4d.

**Comment by jeremyckahn** [comment]
@gemini-code-assist Can you spot any potential regressions in this refactor? The tests all pass, but I'm curious to see if you can spot any gaps. Be thorough in your analysis and take special case to consider edge cases that may no longer be getting handled appropriately.

**Comment by jeremyckahn** [comment]
@gemini-code-assist I've made some updates to fix some bugs I found. Take another look at the refactor and try to spot any remaining gaps between the latest implementation and the original. Be thorough with considering edge cases or substantive differences in the behavior of `incrementDay`.
